### PR TITLE
Add new compiler option --rootDir

### LINKED ISSUE #2644: Empty parent directories are not preserved with --outDir
I have a source structure in which there is a single top-level directory, and I expect the same structure to be produced in the outDir, instead it is dropped. It seems that tsc trims singleton parent directories.
That behavior is okay, but I would like to be able to explicitly say what the baseDir is, and have the output structure match the input, relative to the baseDir.

Here is a session which demonstrates the problem:
```
alexeagle-macbookpro2:tmp alexeagle$ mkdir -p dir1/dir2/dir3
alexeagle-macbookpro2:tmp alexeagle$ cat > dir1/dir2/dir3/test3.ts
console.log("dir3");
alexeagle-macbookpro2:tmp alexeagle$ mkdir out
alexeagle-macbookpro2:tmp alexeagle$ tsc --outDir out dir1/dir2/dir3/test3.ts
alexeagle-macbookpro2:tmp alexeagle$ ls out/
test3.js
alexeagle-macbookpro2:tmp alexeagle$ cat > dir1/dir2/test2.ts
console.log('dir2');
alexeagle-macbookpro2:tmp alexeagle$ tsc --outDir out dir1/dir2/dir3/test3.ts dir1/dir2/test2.ts
alexeagle-macbookpro2:tmp alexeagle$ ls out
dir3	test2.js
```

It's kind of surprising that by adding a file in dir2, the test3.js file has disappeared from the outDir.

### PR DESCRIPTION
**Pull request by mhegazy**
The new option fixes #2644, which is a reactivation of #287, and has been in discussion in #2034.

#### Background
The current behavior of the compiler when using `--outDir` is to duplicate the input structure in the output under the value of `--outDir`. the root of the input is computed as the longest common path of all source (.ts and not .d.ts) files. so an input of `FolderA\FolderB\1.ts` and `FolderA\FolderB\2.ts` would result in computing the commonSourceRoot to `FolderA\FolderB\`. now if a new file `FolderA\3.ts` is added to the input, the commonSourceRoot will pop out to `FolderA\`.

This is generally undesirable and can be confusing. it also does not allow for folders that do not contain .ts files to be replicated in the output. 

#### `--rootDir` flag
The new flag is used to customize the commonSourceRoot instead of automatically calculating it. 

The same rules apply, i.e. all source files (.ts) should be under the rootDir, otherwise the output generation may not be correct.

The name `rootDir` is meant to resemble `outDir` as both flags go together.

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -1968,6 +1968,15 @@
         "category": "Message",
         "code": 6057
     },
+    "Specifies the root directory of input files. Use to control the output directory structure with --outDir.": {
+        "category": "Message",
+        "code": 6058
+    },
+    "File '{0}' is not under rootDir '{1}'. RootDir is expected to contain all source files.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: Use lowercase 'r' for 'RootDir', use single quotes :

"under 'rootDir' '{1}'. 'rootDir' is expected..."

File: src/compiler/program.ts [hunk 2]
```diff
@@ -453,6 +453,66 @@ module ts {
             }
         }
 
+        function computecommonSourceDirectory(sourceFiles: SourceFile[]): string {
+            let commonPathComponents: string[];
+            let currentDirectory = host.getCurrentDirectory();
+            forEach(files, sourceFile => {
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: I think you meant `sourceFiles` here not `files`.

Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Also use a for-of.

File: src/compiler/program.ts [hunk 3]
```diff
@@ -453,6 +453,66 @@ module ts {
             }
         }
 
+        function computecommonSourceDirectory(sourceFiles: SourceFile[]): string {
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: compute***C***ommon

File: src/compiler/program.ts [hunk 4]
```diff
@@ -453,6 +453,66 @@ module ts {
...
+            forEach(files, sourceFile => {
+                // Each file contributes into common source file path
+                if (isDeclarationFile(sourceFile)) {
+                    return;
+                }
+
+                let sourcePathComponents = getNormalizedPathComponents(sourceFile.fileName, currentDirectory);
+                sourcePathComponents.pop(); // FileName is not part of directory
+                if (commonPathComponents) {
+                    for (let i = 0; i < Math.min(commonPathComponents.length, sourcePathComponents.length); i++) {
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Cache the min

File: src/compiler/program.ts [hunk 5]
```diff
@@ -453,6 +453,66 @@ module ts {
...
+            let currentDirectory = host.getCurrentDirectory();
+            forEach(files, sourceFile => {
+                // Each file contributes into common source file path
+                if (isDeclarationFile(sourceFile)) {
+                    return;
+                }
+
+                let sourcePathComponents = getNormalizedPathComponents(sourceFile.fileName, currentDirectory);
+                sourcePathComponents.pop(); // FileName is not part of directory
+                if (commonPathComponents) {
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]:
```TypeScript
if (!commonPathComponents) {
    commonPathComponents = sourcePathComponents;
    continue; // or return if you don't change switch to a for..of
}
```

File: src/compiler/program.ts [hunk 6]
```diff
@@ -453,6 +453,66 @@ module ts {
...
+                    for (let i = 0; i < Math.min(commonPathComponents.length, sourcePathComponents.length); i++) {
+                        if (commonPathComponents[i] !== sourcePathComponents[i]) {
+                            if (i === 0) {
+                                diagnostics.add(createCompilerDiagnostic(Diagnostics.Cannot_find_the_common_subdirectory_path_for_the_input_files));
+                                return;
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Change to continue if it becomes a for..of

File: src/compiler/program.ts [hunk 7]
```diff
@@ -453,6 +453,66 @@ module ts {
...
+                            }
+
+                            // New common path found that is 0 -> i-1
+                            commonPathComponents.length = i;
+                            break;
+                        }
+                    }
+
+                    // If the fileComponent path completely matched and less than already found update the length
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Not sure what you mean.

File: src/compiler/program.ts (hunk 5 above)
Comment (mhegazy) [inline src/compiler/program.ts]: done.

File: src/compiler/program.ts [hunk 8]
```diff
@@ -454,6 +454,66 @@ module ts {
...
+                    // first file
+                    commonPathComponents = sourcePathComponents;
+                    return;
+                }
+            });
+
+            return getNormalizedPathFromPathComponents(commonPathComponents);
+        }
+
+        function checkSourceFilesBelongToPath(soruceFiles: SourceFile[], rootDirectory: string): boolean {
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: soruceFiles -> sourceFiles

File: src/compiler/program.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: It doesn't seem it though

File: src/compiler/program.ts (hunk 8 above)
Comment (basarat) [inline src/compiler/program.ts]: why so soruce?

File: src/compiler/program.ts [hunk 9]
```diff
@@ -454,6 +454,66 @@ module ts {
...
+            let allFilesBelongToPath = true;
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]:
```
                           ~~~~~~~~~~~~~~~~~~~~~
Warning (508, 28) Variable 'allFilesBelongToPath' is declared but never read from.
```

File: src/compiler/program.ts [hunk 10]
```diff
@@ -454,6 +454,66 @@ module ts {
...
+
+                            // New common path found that is 0 -> i-1
+                            commonPathComponents.length = i;
+                            break;
+                        }
+                    }
+
+                    // If the fileComponent path completely matched and less than already found update the length
+                    if (sourcePathComponents.length < commonPathComponents.length) {
+                        commonPathComponents.length = sourcePathComponents.length;
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: You could technically use `n` here.

File: src/compiler/program.ts (hunk 5 above)
Comment (mhegazy) [inline src/compiler/program.ts]: i did add the return.. not sure why i need to move the else clause though

File: src/compiler/program.ts (hunk 9 above)
Comment (mhegazy) [inline src/compiler/program.ts]:
```ts
 return allFilesBelongToPath;
```

Comment (mhegazy) [inline src/compiler/program.ts]: just to report all violations instead of just the first one.

File: src/compiler/program.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: I'm saying instead of having a big `if` statement body, just check perform the `else` clause when applicable

**Comment by DanielRosenwasser** [comment]
:+1:

**Comment by NekR** [comment]
Is this in beta already or will be only in release?

**Comment by mhegazy** [comment]
Should be in 1.5-beta

**Comment by NekR** [comment]
@mhegazy filed here #2974

**Pull request merged**